<a href="https://colab.research.google.com/github/cristiandarioortegayubro/UNI/blob/main/colab/rg_rlm_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://github.com/cristiandarioortegayubro/UNI/blob/main/UNI%2039%20Aniversario.jpg?raw=true)

![](https://scikit-learn.org/stable/_static/scikit-learn-logo-small.png)

# **Regresión**


**La regresión lineal es un método estadístico que trata de modelar la relación entre una variable continua y una o más variables independientes mediante el ajuste de una ecuación lineal. Se llama regresión lineal simple cuando solo hay una variable independiente y regresión lineal múltiple cuando hay más de una. Dependiendo del contexto, a la variable modelada se le conoce como variable dependiente o variable respuesta, y a las variables independientes como regresores, predictores o features.**

# ***Regresión Lineal Multiple***

***Supóngase que el departamento de ventas de una empresa quiere estudiar la influencia que tiene la publicidad a través de distintos canales sobre el número de ventas de un producto. Se dispone de un conjunto de datos que contiene los ingresos (en millones) conseguido por ventas en 200 regiones, así como la cantidad de presupuesto, también en millones, destinado a anuncios por radio, TV y periódicos en cada una de ellas.***


## **Bibliotecas**

### **Instalando bibliotecas**

In [1]:
!pip install plotly --upgrade

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 15.3 MB 5.0 MB/s 
  Attempting uninstall: plotly
    Found existing installation: plotly 5.5.0
    Uninstalling plotly-5.5.0:
      Successfully uninstalled plotly-5.5.0


### **Análisis de datos**

In [2]:
import numpy as np
import pandas as pd

### **Gráficos**

In [3]:
import plotly.express as px

### **Preprocesado y modelado**

#### **Con Scikit-learn**

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

#### **Con Statsmodels**

In [5]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

## **Conjunto de Datos**

In [6]:
url = "https://raw.githubusercontent.com/cristiandarioortegayubro/BA/main/Datasets/publicidad_multiple.csv"

## ***Dataframe***

In [7]:
datos = pd.read_csv(url, index_col = 0)

In [8]:
datos.head(10)

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
1    44.5   39.3  45.1       10.4
2    17.2   45.9  69.3        9.3
3   151.5   41.3  58.5       18.5
4   180.8   10.8  58.4       12.9
5     8.7   48.9  75.0        7.2
6    57.5   32.8  23.5       11.8
7   120.2   19.6  11.6       13.2
8     8.6    2.1   1.0        4.8
9   199.8    2.6  21.2       10.6

## **Relación entre variables**

***El primer paso a la hora de establecer un modelo lineal múltiple es estudiar la relación que existe entre variables. Esta información es crítica a la hora de identificar cuáles pueden ser los mejores predictores para el modelo, y para detectar colinealidad entre predictores. A modo complementario, es recomendable representar la distribución de cada variable mediante histogramas.***

In [9]:
corr_matrix = round(datos.corr(),3)

### **Matriz de correlaciones**

In [10]:
px.imshow(corr_matrix,
          title = "Matriz de correlacion",
          text_auto = True, 
          template = "gridon",
          labels={"color":"Coeficiente"})

### **Gráfico de distribución para cada variable numérica**

In [11]:
datos.columns

Index(['ventas', 'radio', 'tv', 'periodico'], dtype='object')

In [12]:
for i in datos.columns:
  fig = px.histogram(datos, 
                     x = i,
                     template = "gridon",
                     title = i)
  
  fig.update_layout(bargap=0.2)
  fig.show()

## ***Ajuste del modelo***

**Se ajusta un modelo lineal múltiple con el objetivo de predecir las ventas en función de la inversión en los tres canales de publicidad.**


### **División de los datos en train y test**

In [13]:
X = datos[["radio","tv","periodico"]]
y = datos[["ventas"]]

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y,
                                                    train_size = 0.8,
                                                    random_state = 12345)

In [15]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   ventas     200 non-null    float64
 1   radio      200 non-null    float64
 2   tv         200 non-null    float64
 3   periodico  200 non-null    float64
dtypes: float64(4)
memory usage: 7.8 KB


In [16]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 160 entries, 70 to 29
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   radio      160 non-null    float64
 1   tv         160 non-null    float64
 2   periodico  160 non-null    float64
dtypes: float64(3)
memory usage: 5.0 KB


In [17]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40 entries, 52 to 100
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   radio      40 non-null     float64
 1   tv         40 non-null     float64
 2   periodico  40 non-null     float64
dtypes: float64(3)
memory usage: 1.2 KB


In [18]:
y_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40 entries, 52 to 100
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ventas  40 non-null     float64
dtypes: float64(1)
memory usage: 640.0 bytes


In [19]:
y_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 160 entries, 70 to 29
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ventas  160 non-null    float64
dtypes: float64(1)
memory usage: 2.5 KB


### **Creación del modelo**

[statsmodels](https://www.statsmodels.org/stable/index.html)

***Statsmodels es un módulo de Python que proporciona clases y funciones para la estimación de muchos modelos estadísticos diferentes, así como para realizar pruebas estadísticas y exploración de datos estadísticos. Hay disponible una lista extensa de estadísticas de resultados para cada estimador. Los resultados se comparan con los paquetes estadísticos existentes para garantizar que sean correctos. El paquete se publica bajo la licencia BSD modificada (3 cláusulas) de código abierto. La documentación en línea está alojada en statsmodels.org***

In [20]:
X_train = sm.add_constant(X_train, prepend=True)
modelo = sm.OLS(endog = y_train, exog=X_train,)
modelo = modelo.fit()
print(modelo.summary())

                            OLS Regression Results                            
Dep. Variable:                 ventas   R-squared:                       0.861
Model:                            OLS   Adj. R-squared:                  0.859
Method:                 Least Squares   F-statistic:                     323.2
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           1.05e-66
Time:                        20:36:10   Log-Likelihood:                -776.31
No. Observations:                 160   AIC:                             1561.
Df Residuals:                     156   BIC:                             1573.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -46.5085      8.128     -5.722      0.0

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:142: FutureWarning:

In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only



In [21]:
datos.columns

Index(['ventas', 'radio', 'tv', 'periodico'], dtype='object')

In [22]:
modelo.params

const       -46.508452
radio        -3.650996
tv            0.147380
periodico    19.447756
dtype: float64

$$ventas = -\ 46.508452 - 3.650996\ radio + 0.147380\ tv + 19.447756\ periodico$$

In [23]:
print(f"{datos.columns[0]} = {round(modelo.params[0],2)} {round(modelo.params[1],2)}{datos.columns[1]} + {round(modelo.params[2],2)}{datos.columns[2]} + {round(modelo.params[3],2)}{datos.columns[3]}")

ventas = -46.51 -3.65radio + 0.15tv + 19.45periodico


In [24]:
modelo.rsquared

0.861415996555243

***El modelo con todas las variables introducidas como predictores tiene un $𝑅^2$  alto (0.861), es capaz de explicar el 86.1% de la variabilidad observada en las ventas.***

***Acorde al p-value obtenido para el coeficiente parcial de regresión de tv (0.239), esta variable no contribuye de forma significativa al modelo.***

In [25]:
r2 = round(modelo.rsquared, 3)
print(f"R2 = {r2}")

R2 = 0.861


## ***Predicciones***

***Una vez entrenado el modelo, se pueden obtener predicciones para nuevos datos. Los modelos de statsmodels permiten calcular los intervalos de confianza asociados a cada predicción.***



### ***Predicciones con intervalo de confianza***

In [26]:
predicciones = modelo.get_prediction(exog = X_train).summary_frame(alpha=0.05)
predicciones.head(4)

mean   mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
70   203.368619  3.333358     196.784279     209.952960    141.066618   
154  181.240642  3.621162     174.087806     188.393478    118.875997   
194  160.846562  4.895827     151.175895     170.517229     98.143231   
33   218.906802  4.838583     209.349210     228.464394    156.220811   

     obs_ci_upper  
70     265.670621  
154    243.605287  
194    223.549893  
33     281.592793

## ***Interpretación***

***El modelo de regresión lineal múltiple:***

$$ventas = -46.5085 - 3.6510\ radio + 0.1474\ tv + 19.4478\ periodico$$
 
***es capaz de explicar el $86.1$% de la varianza observada en las ventas. Por lo que hay evidencias claras de que el modelo es capaz de explicar la varianza en las ventas mejor de lo esperado por azar. Los test estadísticos para cada variable confirman que tv, radio y periodico están relacionadas con la cantidad de ventas y contribuyen al modelo.***

# Ajuste del modelo a resultado positivo por región (utilidades)

In [27]:
datos.head(3)

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
1    44.5   39.3  45.1       10.4
2    17.2   45.9  69.3        9.3

***El dataframe contiene por cada región, las ventas y la inversión en los canales de publicidad.***

In [28]:
datos.sum()

ventas       29408.5
radio         4652.8
tv            6110.8
periodico     2804.5
dtype: float64

***Se puede observar que las ventas totales superan ampliamente los gastos en canales de publicidad***

In [29]:
datos["utilidad"] = datos.ventas - datos.radio - datos.tv - datos.periodico

In [30]:
datos.head(3)

ventas  radio    tv  periodico  utilidad
0   230.1   37.8  69.2       22.1     101.0
1    44.5   39.3  45.1       10.4     -50.3
2    17.2   45.9  69.3        9.3    -107.3

In [31]:
utilidad = round(datos.utilidad.sum(),2)
print(f"La utilidad total es: {utilidad}")

La utilidad total es: 15840.4


## Filtrando por utilidad positiva

In [32]:
utilidad = datos[(datos.utilidad > 0)]
utilidad = utilidad.copy()
utilidad.reset_index(drop = True)
utilidad.drop(columns = "utilidad", inplace = True)
utilidad

ventas  radio    tv  periodico
0     230.1   37.8  69.2       22.1
3     151.5   41.3  58.5       18.5
4     180.8   10.8  58.4       12.9
7     120.2   19.6  11.6       13.2
8       8.6    2.1   1.0        4.8
..      ...    ...   ...        ...
195    38.2    3.7  13.8        7.6
196    94.2    4.9   8.1        9.7
197   177.0    9.3   6.4       12.8
198   283.6   42.0  66.2       25.5
199   232.1    8.6   8.7       13.4

[153 rows x 4 columns]

## Relación entre variables

In [33]:
corr_matrix = round(utilidad.corr(),3)

### Matriz de correlaciones

In [34]:
px.imshow(corr_matrix,
          title = "Matriz de correlacion",
          text_auto = True, 
          template = "gridon",
          labels={"color":"Coeficiente"})

### Gráfico de distribución para cada variable numérica

In [35]:
for i in utilidad.columns:
  fig = px.histogram(datos, 
                     x = i,
                     template = "gridon",
                     title = i)
  
  fig.update_layout(bargap=0.2)
  fig.show()

## **Ajuste del modelo**

***Se ajusta un modelo lineal múltiple con el objetivo de predecir las ventas en función de la inversión en los tres canales de publicidad.***


### ***División de los datos en train y test***

In [36]:
X = utilidad[["radio","tv","periodico"]]
y = utilidad["ventas"]

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X.values.reshape(-1,3), 
                                                    y.values.reshape(-1,1),
                                                    train_size = 0.8,
                                                    random_state = 12345,
                                                    shuffle = True)

### **Creación del modelo**

In [38]:
X_train = sm.add_constant(X_train, prepend=True)
modelo = sm.OLS(endog=y_train, exog=X_train,)
modelo = modelo.fit()
print(modelo.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.864
Model:                            OLS   Adj. R-squared:                  0.860
Method:                 Least Squares   F-statistic:                     249.0
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           7.39e-51
Time:                        20:36:10   Log-Likelihood:                -565.64
No. Observations:                 122   AIC:                             1139.
Df Residuals:                     118   BIC:                             1151.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -112.1096     11.028    -10.166      0.0

In [39]:
modelo.params

array([-1.12109579e+02, -6.81085614e+00, -6.60987929e-02,  2.83210076e+01])

$$ventas = -\ 112.109579 - 6.810856\ radio - 0.066099\ tv + 28.321008\ periodico$$

In [40]:
r2 = round(modelo.rsquared, 3)
print(f"R2 = {r2}")

R2 = 0.864


El filtrado en el conjunto de datos para regiones con utilidad positiva incrementa el valor $R^2$ de 0.861 a 0.864

# **Primer escenario**

**Eliminando la variable no significativa (tv)**

In [41]:
modelo.params

array([-1.12109579e+02, -6.81085614e+00, -6.60987929e-02,  2.83210076e+01])

In [42]:
escenario01 = utilidad.copy(deep=True)
escenario01.drop(columns="tv", inplace=True)

In [43]:
escenario01.head()

ventas  radio  periodico
0   230.1   37.8       22.1
3   151.5   41.3       18.5
4   180.8   10.8       12.9
7   120.2   19.6       13.2
8     8.6    2.1        4.8

## **Relación entre variables**

In [44]:
corr_matrix = round(escenario01.corr(),3)

### **Matriz de correlaciones**

In [45]:
px.imshow(corr_matrix,
          title = "Matriz de correlacion",
          text_auto = True, 
          template = "gridon",
          labels={"color":"Coeficiente"})

### **Gráfico de distribución para cada variable numérica**

In [46]:
for i in escenario01.columns:
  fig = px.histogram(datos, 
                     x = i,
                     template = "gridon",
                     title = i)
  
  fig.update_layout(bargap=0.2)
  fig.show()

## **Ajuste del modelo**

***Se ajusta un modelo lineal múltiple con el objetivo de predecir las ventas en función de la inversión en los dos canales de publicidad.***


### **División de los datos en train y test**

In [47]:
X = escenario01.drop(columns=["ventas"])
y = escenario01["ventas"]

In [48]:
X_train, X_test, y_train, y_test = train_test_split(X.values.reshape(-1,2), 
                                                    y.values.reshape(-1,1),
                                                    train_size = 0.8,
                                                    random_state = 12345,
                                                    shuffle = True)

### Creación del modelo

In [49]:
X_train = sm.add_constant(X_train, prepend=True)
modelo01 = sm.OLS(endog=y_train, exog=X_train,)
modelo01 = modelo01.fit()
print(modelo01.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.863
Model:                            OLS   Adj. R-squared:                  0.861
Method:                 Least Squares   F-statistic:                     375.5
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           3.91e-52
Time:                        20:36:11   Log-Likelihood:                -565.80
No. Observations:                 122   AIC:                             1138.
Df Residuals:                     119   BIC:                             1146.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -112.0272     10.994    -10.189      0.0

In [50]:
modelo01.params

array([-112.02721393,   -6.79169465,   28.16636573])

$$ventas = -\ 112.027214 - 6.791695\ radio + 28.166366\ periodico$$

In [51]:
r2 = round(modelo01.rsquared, 3)
print(f"R2 = {r2}")

R2 = 0.863


# **Segundo escenario**

**Distribuyendo la variable no significativa (tv)**

In [52]:
modelo.params

array([-1.12109579e+02, -6.81085614e+00, -6.60987929e-02,  2.83210076e+01])

In [53]:
escenario02 = utilidad.copy(deep=True)

In [54]:
escenario02["radioperiodico"] = escenario02.radio + escenario02.periodico
escenario02["coef_radio"] = escenario02.radio / escenario02.radioperiodico
escenario02["coef_periodico"] = escenario02.periodico / escenario02.radioperiodico
escenario02["prorr_coef_radio"] = escenario02.coef_radio * escenario02.tv
escenario02["prorr_coef_periodico"] = escenario02.coef_periodico * escenario02.tv
escenario02["radio"] = round((escenario02.radio + escenario02.prorr_coef_radio),2)
escenario02["periodico"] = round((escenario02.periodico + escenario02.prorr_coef_periodico),2)

In [55]:
escenario02.drop(columns=["radioperiodico","coef_radio","coef_periodico",
                          "prorr_coef_radio","prorr_coef_periodico","tv"], inplace=True)

In [56]:
escenario02.head()

ventas  radio  periodico
0   230.1  81.47      47.63
3   151.5  81.70      36.60
4   180.8  37.41      44.69
7   120.2  26.53      17.87
8     8.6   2.40       5.50

## Comprobación

In [57]:
utilidad.head()

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
3   151.5   41.3  58.5       18.5
4   180.8   10.8  58.4       12.9
7   120.2   19.6  11.6       13.2
8     8.6    2.1   1.0        4.8

In [58]:
utilidad.radio[0] + utilidad.tv[0] + utilidad.periodico[0]

129.1

In [59]:
utilidad.radio + utilidad.tv + utilidad.periodico

0      129.1
3      118.3
4       82.1
7       44.4
8        7.9
       ...  
195     25.1
196     22.7
197     28.5
198    133.7
199     30.7
Length: 153, dtype: float64

In [60]:
escenario02.head()

ventas  radio  periodico
0   230.1  81.47      47.63
3   151.5  81.70      36.60
4   180.8  37.41      44.69
7   120.2  26.53      17.87
8     8.6   2.40       5.50

In [61]:
escenario02.radio[0] + escenario02.periodico[0]

129.1

In [62]:
escenario02.radio + escenario02.periodico

0      129.1
3      118.3
4       82.1
7       44.4
8        7.9
       ...  
195     25.1
196     22.7
197     28.5
198    133.7
199     30.7
Length: 153, dtype: float64

## Relación entre variables

In [63]:
corr_matrix = round(escenario02.corr(),2)

### **Matriz de correlaciones**

In [64]:
px.imshow(corr_matrix,
          title = "Matriz de correlacion",
          text_auto = True, 
          template = "gridon",
          labels={"color":"Coeficiente"})

### Gráfico de distribución para cada variable numérica

In [65]:
for i in escenario02.columns:
  fig = px.histogram(datos, 
                     x = i,
                     template = "gridon",
                     title = i)
  
  fig.update_layout(bargap=0.2)
  fig.show()

## Ajuste del modelo

In [66]:
X = escenario02.drop(columns=["ventas"])
y = escenario02["ventas"]

In [67]:
X_train, X_test, y_train, y_test = train_test_split(X.values.reshape(-1,2), 
                                                    y.values.reshape(-1,1),
                                                    train_size = 0.8,
                                                    random_state = 12345,
                                                    shuffle = True)

### **Creación del modelo**

In [68]:
X_train = sm.add_constant(X_train, prepend=True)
modelo02 = sm.OLS(endog=y_train, exog=X_train,)
modelo02 = modelo02.fit()
print(modelo02.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.261
Model:                            OLS   Adj. R-squared:                  0.248
Method:                 Least Squares   F-statistic:                     20.97
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           1.58e-08
Time:                        20:36:11   Log-Likelihood:                -668.74
No. Observations:                 122   AIC:                             1343.
Df Residuals:                     119   BIC:                             1352.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         92.9522     14.958      6.214      0.0

In [69]:
modelo02.params

array([92.95215472,  0.35717866,  2.58715611])

$$ventas = 92.952155 + 0.357179\ radio + 2.587156\ periodico$$

In [70]:
r2 = round(modelo02.rsquared, 3)
print(f"R2 = {r2}")

R2 = 0.261


# **Tercer escenario**

**Sumando la variable no significativa (tv) a radio**

In [71]:
escenario03 = utilidad.copy(deep=True)

In [72]:
escenario03["radio"] = round((escenario03.radio + escenario03.tv),2)

In [73]:
escenario03.drop(columns="tv", inplace=True)

## Comprobación

In [74]:
utilidad.head()

ventas  radio    tv  periodico
0   230.1   37.8  69.2       22.1
3   151.5   41.3  58.5       18.5
4   180.8   10.8  58.4       12.9
7   120.2   19.6  11.6       13.2
8     8.6    2.1   1.0        4.8

In [75]:
utilidad.radio[0] + utilidad.tv[0] + utilidad.periodico[0]

129.1

In [76]:
escenario03.head()

ventas  radio  periodico
0   230.1  107.0       22.1
3   151.5   99.8       18.5
4   180.8   69.2       12.9
7   120.2   31.2       13.2
8     8.6    3.1        4.8

In [77]:
escenario03.radio[0] + escenario03.periodico[0]

129.1

## Relación entre variables

In [78]:
corr_matrix = round(escenario03.corr(),3)

### Matriz de correlaciones

In [79]:
px.imshow(corr_matrix,
          title = "Matriz de correlacion",
          text_auto = True, 
          template = "gridon",
          labels={"color":"Coeficiente"})

### Gráfico de distribución para cada variable numérica

In [80]:
for i in escenario03.columns:
  fig = px.histogram(datos, 
                     x = i,
                     template = "gridon",
                     title = i)
  
  fig.update_layout(bargap=0.2)
  fig.show()

## Ajuste del modelo

In [81]:
X = escenario03.drop(columns=["ventas"])
y = escenario03["ventas"]

In [82]:
X_train, X_test, y_train, y_test = train_test_split(X.values.reshape(-1,2), 
                                                    y.values.reshape(-1,1),
                                                    train_size = 0.8,
                                                    random_state = 12345,
                                                    shuffle = True)

### Creación del modelo

In [83]:
X_train = sm.add_constant(X_train, prepend=True)
modelo03 = sm.OLS(endog=y_train, exog=X_train,)
modelo03 = modelo03.fit()
print(modelo03.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.510
Model:                            OLS   Adj. R-squared:                  0.501
Method:                 Least Squares   F-statistic:                     61.81
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           3.91e-19
Time:                        20:36:12   Log-Likelihood:                -643.70
No. Observations:                 122   AIC:                             1293.
Df Residuals:                     119   BIC:                             1302.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         20.7040     15.107      1.371      0.1

In [84]:
modelo03.params

array([20.70395099, -0.53786837, 12.0923078 ])

$$ventas = 20.703951 - 0.537868\ radio + 12.092308\ periodico$$

In [85]:
r2 = round(modelo03.rsquared, 3)
print(f"R2 = {r2}")

R2 = 0.51
